torchao/dtypes/nf4tensor.py

```diff
@@ -616,9 +972,64 @@ def function_to_dtype(*args, **kwargs):
     if isinstance(args[0], NF4Tensor) and isinstance(args[1], torch.dtype):
         # Tensor.to(dtype, non_blocking, copy, memory_format)
         return args[0].get_original_weight().to(*args[1:], **kwargs)
+    elif isinstance(args[0], NF4Tensor) and (
+        isinstance(args[1], torch.device) or (
+            isinstance(args[1], str) and (
+                args[1] == "cpu" or args[1].startswith("cuda")
+            )
+        )
+    ) and len(args) == 2:
+        # Tensor.to(device, non_blocking)
```

Comment:
For cpu offloading, FSDP2 call `NF4Tensor.to(device, non_blocking=True)`. I have to implement this in `__torch_function__`. In `__torch_dispatch__`, `aten._to_copy` contains default dtype=bf16/fp16/fp32 and we do not want dequant

Reply:
Does this mean that if we tried to use `__torch_dispatch__`, we would not be able to tell that it is simply `.to(device, non_blocking=True)` without a dtype argument/dtype change?

What is the story for dequantization? Namely, what is the outer `NF4Tensor`'s dtype, and what happens when you call `.to(dtype)` with that same dtype? (e.g. if `NF4Tensor.dtype == torch.bfloat16`, what if you call `NF4Tensor.to(torch.bfloat16)`?)

Reply:
> if we tried to use __torch_dispatch__, we would not be able to tell that it is simply .to(device, non_blocking=True) without a dtype argument/dtype change?

exactly

> what is the outer NF4Tensor's dtype

say we construct NF4Tensor from fp32 tensor, then NF4Tensor.dtype = fp32. it's always the original tensor dtype

> what happens when you call .to(dtype) with that same dtype

it will dequant because we implemented `.to(dtype)` in  `__torch_function__`

Reply:
dequant through `tensor.get_original_weight()` in `__torch_function__`